service/history/shard/context_util.go

```diff
@@ -0,0 +1,164 @@
+				logger.Warn(
+					"Shard queue lag exceeds warn threshold.",
+					tag.ShardQueueAcks(category.Name(), minTaskKey.TaskID),
+				)
+			}
+			metricsHandler.Histogram(
+				metrics.ShardInfoImmediateQueueLagHistogram.GetMetricName(),
+				metrics.ShardInfoImmediateQueueLagHistogram.GetMetricUnit(),
+			).Record(lag, metrics.TaskCategoryTag(category.Name()))
+
+		case tasks.CategoryTypeScheduled:
+			minTaskKey := getMinTaskKey(queueState)
+			if minTaskKey == nil {
+				continue Loop
+			}
+			lag := scheduledTaskMaxReadLevel.Sub(minTaskKey.FireTime)
+			if lag > logWarnScheduledTaskLag {
+				logger.Warn(
+					"Shard queue lag exceeds warn threshold.",
+					tag.ShardQueueAcks(category.Name(), minTaskKey.FireTime),
+				)
+			}
+			metricsHandler.Timer(
+				metrics.ShardInfoScheduledQueueLagTimer.GetMetricName(),
+			).Record(lag, metrics.TaskCategoryTag(category.Name()))
+		default:
+			logger.Error("Unknown task category type", tag.NewStringTag("task-category", category.Type().String()))
+		}
+	}
+}
+
+func convertPersistenceAckLevelToTaskKey(
+	categoryType tasks.CategoryType,
+	ackLevel int64,
+) tasks.Key {
+	if categoryType == tasks.CategoryTypeImmediate {
+		return tasks.NewImmediateKey(ackLevel)
+	}
+	return tasks.NewKey(timestamp.UnixOrZeroTime(ackLevel), 0)
+}
+
+func convertTaskKeyToPersistenceAckLevel(
+	categoryType tasks.CategoryType,
+	taskKey tasks.Key,
+) int64 {
+	if categoryType == tasks.CategoryTypeImmediate {
+		return taskKey.TaskID
+	}
+	return taskKey.FireTime.UnixNano()
+}
+
+func ConvertFromPersistenceTaskKey(
+	key *persistencespb.TaskKey,
+) tasks.Key {
+	return tasks.NewKey(
+		timestamp.TimeValue(key.FireTime),
+		key.TaskId,
+	)
+}
+
+func ConvertToPersistenceTaskKey(
+	key tasks.Key,
+) *persistencespb.TaskKey {
+	return &persistencespb.TaskKey{
+		FireTime: timestamp.TimePtr(key.FireTime),
+		TaskId:   key.TaskID,
+	}
+}
+
+// ReplicationReaderIDFromClusterShardID convert from cluster ID & shard ID to reader ID
+// NOTE: cluster metadata guarantee
+//  1. initial failover version <= int32 max
+//  2. failover increment <= int32 max
+//  3. initial failover version == cluster ID
+func ReplicationReaderIDFromClusterShardID(
+	clusterID int64,
+	shardID int32,
+) int64 {
+	return clusterID<<32 + int64(shardID)
+}
+
+// ReplicationReaderIDToClusterShardID convert from reader ID to cluster ID & shard ID
+// NOTE: see ReplicationReaderIDFromClusterShardID
+func ReplicationReaderIDToClusterShardID(
+	readerID int64,
+) (int64, int32) {
+	return readerID >> 32, int32(readerID & 0xffffffff)
+}
+
+func getMinTaskKey(
+	queueState *persistencespb.QueueState,
+) *tasks.Key {
+	var minTaskKey *tasks.Key
+	if queueState.ExclusiveReaderHighWatermark != nil {
+		taskKey := ConvertFromPersistenceTaskKey(queueState.ExclusiveReaderHighWatermark)
+		minTaskKey = &taskKey
+	}
+	for _, readerState := range queueState.ReaderStates {
+		taskKey := ConvertFromPersistenceTaskKey(readerState.Scopes[0].Range.InclusiveMin)
```

Comment:
I see a couple of places where [0] is directly access, I do understand right now scopes must be non-empty, but someone may break it in the future... :( 

I am thinking if is it possible to trim those empty slices from map on loading shard info & Set/UpdateQueueState? Maybe you have a better idea?

Reply:
sure, will send another PR

Reply:
actually, does https://github.com/temporalio/temporal/pull/4190/files/167dc0be71db8e4010720c04f8a3021133fa01b4#diff-14665987ac1f72fe1eec59b480b6463c209c117352308819c88585805ec66fc8R227 address your concern?

Reply:
this only performs trimming for replication? I think we'll want to protect all categories.